src/main/java/conversion/BaseServlet.java

```diff
@@ -69,118 +103,364 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+     * @param response
+     * @see BaseServlet#allowCrossOrigin(HttpServletResponse)
+     */
     @Override
     protected void doOptions(HttpServletRequest request, HttpServletResponse response) {
         allowCrossOrigin(response);
     }
 
+    /**
+     * Allow cross origin requests according to the CORS standard.
+     *
+     * @param response
+     */
     private void allowCrossOrigin(final HttpServletResponse response) {
         response.addHeader("Access-Control-Allow-Origin", "*");
         response.addHeader("Access-Control-Allow-Methods", "GET, PUT, POST, OPTIONS, DELETE");
         response.addHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Access-Control-Allow-Origin");
     }
 
+    /**
+     * A post request to the server. This method deals with the initial contact
+     * with the client. It looks for a file within the http request itself, it
+     * then falls back to looking for a file at a url that has passed in the
+     * parameters.
+     * <p>
+     * If a file or url is not present a 400 error is returned. If the file
+     * cannot be parsed from the http request then a 500 error is returned. If a
+     * file cannot be downloaded from a passed url then the individual state is
+     * set to "error" and the conversion does not happen.
+     * <p>
+     * If the file can be found and has been downloaded then an input dir and
+     * and output dir for that file are created (overwritten if they already
+     * exist).
+     * <p>
+     * The convert() method is then started in a separate thread.
+     *
+     * @param request
+     * @param response
+     * @see BaseServlet#convert(Individual, Map, String, String, String, String,
+     * String, String)
+     */
     @Override
     protected void doPost(final HttpServletRequest request, final HttpServletResponse response) {
-
         try {
             allowCrossOrigin(response);
 
             final String uuidStr = UUID.randomUUID().toString();
             final Individual individual = new Individual(uuidStr);
+            final String contextUrl = getContextURL(request);
 
             imap.entrySet().removeIf(entry -> entry.getValue().timestamp < new Date().getTime() - 86400000); // 24 hours
 
             imap.put(uuidStr, individual);
 
             individual.isAlive = true;
 
-            final Part filePart = request.getPart("file");
-            if (filePart == null) {
-                imap.remove(uuidStr);
-                doError(response, "Missing file", 400);
-                return;
-            }
+            boolean isUrlDownload = false;
 
-            final byte[] fileBytes = new byte[(int) filePart.getSize()];
-            final InputStream fileContent = filePart.getInputStream();
-            fileContent.read(fileBytes);
-            fileContent.close();
+            // Attempt to get the filename from parameters (null if not in params).
+            String fileName = request.getParameter("filename");
```

Comment:
Adding a parameter filename is the only way I could think of handing the filename to the server if the client uses a conversion url. If there is no filename then the server tries to extract a filename from the url. 
It made sense (in my head) to keep the filename option for normal file upload conversions.

Reply:
My concern is that it adds complexity to all the clients which will need to handle the optional filename parameter.

One of our main goals is to keep the examples as simple as possible and therefore easy to understand and extend (hence why I want to think carefully about any added complexity).

What if we used a fallback name such as document.pdf?

Reply:
Those points make sense. document.pdf sounds like a good idea, especially since there won't be any file name conflicts with how the files are stored.
I will modify the servlet and then update the clients again to stop taking in filenames.